tests/core/math/test_projection.h

```diff
@@ -0,0 +1,637 @@
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+}
+
+TEST_CASE("[Projection] create_perspective") {
+	SUBCASE("Flipped y") {
+		Projection expected(
+				Vector4(76.39, 0, 0, 0),
+				Vector4(0, 99.307, 0, 0),
+				Vector4(0, 0, -3, -1),
+				Vector4(0, 0, -4, 0));
+		Projection actual = Projection().create_perspective(1.5, 1.3, 1.0, 2.0, true);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+
+	SUBCASE("Not flipped y") {
+		Projection expected(
+				Vector4(58.7615, 0, 0, 0),
+				Vector4(0, 76.39, 0, 0),
+				Vector4(0, 0, -3, -1),
+				Vector4(0, 0, -4, 0));
+		Projection actual = Projection().create_perspective(1.5, 1.3, 1.0, 2.0, false);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+}
+
+TEST_CASE("[Projection] create_perspective_hmd") {
+	SUBCASE("No y flipping, eye 1") {
+		Projection expected(
+				Vector4(57.4362, 0, 0, 0),
+				Vector4(0, 76.39, 0, 0),
+				Vector4(41.6412, 0, -3, -1),
+				Vector4(416.412, 0, -4, 0));
+		Projection actual = Projection().create_perspective_hmd(1.5, 1.33, 1.0, 2.0, false, 1, 14.5, 10.0);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+
+	SUBCASE("No y flipping, eye 2") {
+		Projection expected(
+				Vector4(57.4362, 0, 0, 0),
+				Vector4(0, 76.39, 0, 0),
+				Vector4(-41.6412, 0, -3, -1),
+				Vector4(-416.412, 0, -4, 0));
+		Projection actual = Projection().create_perspective_hmd(1.5, 1.33, 1.0, 2.0, false, 2, 14.5, 10.0);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+
+	SUBCASE("Y flipping, eye 1") {
+		Projection expected(
+				Vector4(76.3899, 0, 0, 0),
+				Vector4(0, 101.599, 0, 0),
+				Vector4(55.3827, 0, -3, -1),
+				Vector4(553.827, 0, -4, 0));
+		Projection actual = Projection().create_perspective_hmd(1.5, 1.33, 1.0, 2.0, true, 1, 14.5, 10.0);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+
+	SUBCASE("Y flipping, eye 2") {
+		Projection expected(
+				Vector4(76.3899, 0, 0, 0),
+				Vector4(0, 101.599, 0, 0),
+				Vector4(-55.3827, 0, -3, -1),
+				Vector4(-553.827, 0, -4, 0));
+		Projection actual = Projection().create_perspective_hmd(1.5, 1.33, 1.0, 2.0, true, 2, 14.5, 10.0);
+
+		for (int i = 0; i < 4; i++) {
+			for (int j = 0; j < 4; j++) {
+				CHECK(expected[i][j] == doctest::Approx(actual[i][j]));
+			}
+		}
+	}
+}
+
+TEST_CASE("[Projection] Determinant") {
+	SUBCASE("Determinant calculation of identity matrix should return 1.0f") {
+		float expected = 1.0;
```

Comment:
```suggestion
		real_t expected = 1.0;
```
This is based on the build precision and should be specific, alternatively
```suggestion
		double expected = 1.0;
```
But not float